### ISSUE TITLE
get_dropout_state, seed and reproducibility

### ISSUE THREAD

**Issue by cyyever** [pytorch/pytorch#74177]
### 🐛 Describe the bug

I am debugging a reproducibility problem in cudnn, specifically, in aten/src/ATen/native/cudnn/rnn.cpp, function get_dropout_state has cached dropout_state, and the cache is not refreshed when the cuda state is reseeded.
I have applied the simple workaround to recreate dropout state and my reproducibility issue has disappeared. 
```
DropoutState& get_dropout_state(double dropout_p, bool train, TensorOptions options) {
  // Each state is slightly over 2MB and initialized lazily, so it's fine to cache them.
  static std::vector<DropoutState> dropout_state_cache { static_cast<size_t>(cuda::getNumGPUs()) };
  static std::mutex state_cache_mut;

  AT_ASSERT(options.device().is_cuda());
  int device = options.device().index();

  std::unique_lock<std::mutex> lock {state_cache_mut};
  auto& state = dropout_state_cache.at(device);
  if (train && dropout_p > 0 && (!state.buffer.defined() || getenv("reseed_dropout"))) {
    std::unique_lock<std::mutex> lock {state.mutex};
    int64_t seed = at::empty({}, at::kLong).random_().item<int64_t>();
    state.buffer = at::_cudnn_init_dropout_state(
        dropout_p, train, seed, options.dtype(at::kByte));
    // NB: CUDA binds the event to a device at creation time, so we can initialize it
    // only now, when we know we're on the correct device.
    if( getenv("reseed_dropout")) {
      TORCH_WARN("reseed dropout");
      unsetenv("reseed_dropout");
    } else {
      state.event.emplace();
    }
  }
  return state;
}

```
Is there a better way to deal with it?

### Versions

I am using the git master of pytorch.

cc @ezyang @gchanan @zou3519 @csarofeen @ptrblck @xwang233
